src/cargo/ops/cargo_clean.rs

```diff
@@ -136,9 +136,13 @@ pub fn clean(ws: &Workspace<'_>, opts: &CleanOptions<'_>) -> CargoResult<()> {
     for pkg in packages {
         let pkg_dir = format!("{}-*", pkg.name());
 
-        // Clean fingerprints.
         for (_, layout) in &layouts_with_host {
+            // Clean fingerprints.
             rm_rf_glob(&layout.fingerprint().join(&pkg_dir), config)?;
+            // Clean target/doc/pkg.
+            rm_rf(&layout.doc().join(pkg.name()), config)?;
```

Comment:
This isn't using the correct name here. The directory names are the crate names of the targets.  I'm uncertain about the best way to handle this, but specifically:

* Non-workspace members can only have library targets documented, so that would be something like `pkg.library().map(|t| t.crate_name())`.  I think it could be dangerous to blindly remove directories for other targets like "examples" because there could be name collisions (for example, if something has an example called "serde", it would delete the serde directory, which is probably not what you want).
* For workspace members, it is possible to document binaries (and now examples due to #9808).  So to clean a package, it would need to delete all documentable targets. Unfortunately this runs afoul with the "serde" example, too, so maybe just deleting libraries is sufficient? I'm still not entirely clear on the motivation for cleaning these, so I'm uncertain if it is important or not.

Reply:
>  I think it could be dangerous to blindly remove directories for other targets like "examples" because there could be name collisions (for example, if something has an example called "serde", it would delete the serde directory, which is probably not what you want).

I'm sorry, I don't quite understand your question. Could you explain it again?

> For workspace members, it is possible to document binaries (and now examples due to Adding the cargo doc --examples subcommand #9808). So to clean a package, it would need to delete all documentable targets. Unfortunately this runs afoul with the "serde" example, too, so maybe just deleting libraries is sufficient? I'm still not entirely clear on the motivation for cleaning these, so I'm uncertain if it is important or not.

I'm sorry. I really didn't think of the question in advance. If the example name in the ../examples/ folder is the same as another crate name, this should be taken into account.

Reply:
Yea, both points are related to one another.

The doc directories are based on a target's crate name. So, for example, if I have a package named "foo" with a "foo" library and a "foo-cli" binary, there would be two directories:

* `doc/foo`
* `doc/foo_cli`

To clean this properly, it would need to iterate over all documentable targets and get their crate names. Don't forget that the package name, the target name, and the crate name can all be different. That is, package "abc-rs" can have a library named "my-abc-lib" which has a crate name of "my_abc_lib".

However, I'd be concerned about indiscriminately iterating over a package's targets because of name collisions. Imagine my "foo" library had a dependency on "serde", and a dependency on another package named "bar". I run `cargo doc` and also get the documentation for `doc/foo`, `doc/serde`, and `doc/bar`. Now, imagine `bar` had an binary coincidentally named `serde`, if a user runs `cargo clean -p bar`, we don't want to blow away the serde directory just because `bar` had a `serde` binary.

So, I'm thinking, to avoid deleting the wrong things, one step would be to only look for library targets of non-workspace dependencies. That should limit the collateral damage.

Then, I'm wondering about limiting the collateral damage of workspace members, too. Perhaps including examples is overkill, and perhaps it should only delete library crate names? Or maybe libraries + binaries? Or maybe it is fine to collect all targets with `documented()` is `true`. That's where I'm uncertain, since I'm not really sure I understand the need for this functionality. 

I think I would lean towards just deleting the entries from `pkg.targets().iter().filter(|t| t.documented())` for workspace members, and just the lib for non-members.

Reply:
I need some time to think, please give me some time. i am sorry, huss.

Reply:
No worries, feel free to take your time, there's no rush.  And let me know if you have any questions or if anything wasn't clear.

Reply:
> What does workspace member mean here?

Essentially, yea, it is the `members` list. It's a little more complicated, since additional members can be inferred.

> My understanding of this function is:
> If you execute `cargo clean -p foo`, you need to delete all `doc/foo`,`doc/example_name`,`doc/dependence_name`.
> And these documents are in `pkg.targets().iter().filter(|t| t.documented())`.

I think my question was: *why* does someone want to delete directories from the `doc` directory? What is the drawback of leaving them there?  What is the use case for needing to delete them? Is it OK for it to only partially delete a package's data?  

> The binary files of `serde` and `bar` seem to have nothing to do with each other. In other words, the binary files under the `bar` directory will not form a document.
> 
> Therefore, running `cargo clean -p bar` only needs to delete `doc/bar`, and does not need to consider the conflict between the `serde` binary file under `bar` and the dependent `serde` of `foo`. Is my understanding correct?

I think what I'm trying to get at is that when you document a package, it documents the libraries and its binaries.  If you have a package `mypackage` with binaries `bar` and `baz`, there will be three directories in the `doc` directory.  To properly clean `mypackage`, it will need to iterate over all documentable targets and delete those directories.  My concern is that could be too aggressive in some cases, as it could delete too much.  

When `cargo doc` is run, it will document all documentable targets for workspace members, and all libraries for non-workspace members.

So my thought is that it would look for `pkg.targets().iter().filter(|t| t.documented())` for workspace members, and `pkg.targets().iter().find(|t| t.is_lib())` for non-members maybe?